cclib/method/ddec.py

```diff
@@ -271,59 +281,105 @@ def calculate(self, indices=None, fupdate=0.05):
             self.radial_grid_r.append(r)
 
         # Carry out step 3 of DDEC6 algorithm [Determine conditioned charge density and tau]
-        self.logger.info("Conditioning charge densities.")
+        self.logger.info("Conditioning charge densities. (Step 3/7)")
         self.condition_densities()
 
-        # STEP 4-7
-        # Optimize weights on each grid point
-        kappa = [0.0 for x in self.data.atomnos]
+        # Steps 4 through 7 contain similar routines. Comments the precede each step explain the
+        # differences among them.
+
+        # * STEP 4 *
+        # In step 4, calculate w, u, g, and h but always skip kappa updates.
+        self._kappa = [0.0] * self.data.atomnos
         update_kappa = False
+        # N_A is assigned number of electrons determined using equation 72.
         self.N_A = []
-        steps = 4
-        while steps < 5:  # change this to 8 after implementing kappa routine
-            self.logger.info("Optimizing charge weights (Step {}).".format(steps))
+        # u_A is determined using equation 77.
+        self.u_A = []
+
+        self.logger.info("Optimizing grid weights. (Step 4/7)")
+        self.N_A.append(self._calculate_w_and_u())
+
+        # Calculate G_A and H_A to update w_A based on S4.3 in doi: 10.1039/c6ra04656h
```

Comment:
Is this comment about the two lines below? Does "calculate G_A" mean the same thing as "reshape G" and I don't see anything related to w_A.

Reply:
I was trying to say that G_A and H_A are calculated here to update w_A, since G_A and H_A do not serve any other purpose than to make sure that w_A does not diverge in the numerical algorithm. I've rephrased this a bit. Thanks.